File: completable-reactor-graph-viewer/src/test/java/ru/fix/completable/reactor/graph/viewer/code/CodeUpdaterTest.java
package ru.fix.completable.reactor.graph.viewer.code;

import org.apache.commons.io.IOUtils;
import org.junit.Assert;
import org.junit.BeforeClass;
import org.junit.Test;
import ru.fix.completable.reactor.api.ReactorGraphModel;
import ru.fix.completable.reactor.graph.viewer.GraphViewer;

import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * @author Kamil Asfandiyarov
 */
public class CodeUpdaterTest {

    final CodeUpdater updater = new CodeUpdater();

    static final List<GraphViewer.CoordinateItem> coordinateItems = new ArrayList<>();

    @BeforeClass
    public static void beforeClass() {
        coordinateItems.add(new GraphViewer.CoordinateItem(
                GraphViewer.CoordinateItem.Type.START_POINT,
                null,
                500,
                -200));

        coordinateItems.add(new GraphViewer.CoordinateItem(
                GraphViewer.CoordinateItem.Type.PROCESSOR,
                new ReactorGraphModel.Identity()
                        .setClassName("SmsRequestConstructProcessor")
                        .setId(4),
                910,
                901));

        coordinateItems.add(new GraphViewer.CoordinateItem(
                GraphViewer.CoordinateItem.Type.PROCESSOR,
                new ReactorGraphModel.Identity()
                        .setClassName("SmsRequestConstructProcessor")
                        .setId(5),
                1170,
                1270));

        coordinateItems.add(new GraphViewer.CoordinateItem(
                GraphViewer.CoordinateItem.Type.PROCESSOR,
                new ReactorGraphModel.Identity()
                        .setClassName("ServiceFetchProcessor")
                        .setId(0),
                320,
                1141));

        coordinateItems.add(new GraphViewer.CoordinateItem(
                GraphViewer.CoordinateItem.Type.MERGE_POINT,
                new ReactorGraphModel.Identity()
                        .setClassName("ServiceFetchProcessor")
                        .setId(5),
                1210,
                401));

        coordinateItems.add(new GraphViewer.CoordinateItem(
                GraphViewer.CoordinateItem.Type.END_POINT,
                new ReactorGraphModel.Identity()
                        .setClassName("SmsRequestConstructProcessor")
                        .setId(6),

                1500,
                511));
    }

    @Test
    public void update_code_block() throws Exception {
        String input = IOUtils.toString(getClass().getResource("/code-block-1.txt").toURI(), StandardCharsets.UTF_8);
        String expectedOutput = IOUtils.toString(getClass().getResource("/code-block-1-result.txt").toURI(), StandardCharsets.UTF_8);

        String output = updater.updateCoordinates(input, coordinateItems);
        Assert.assertEquals(normalize(expectedOutput), normalize(output));
    }

    @Test
    public void code_after_code_Block_should_not_be_changed() throws Exception {
        String input = IOUtils.toString(getClass().getResource("/code-block-2.txt").toURI(), StandardCharsets.UTF_8);
        String expectedOutput = IOUtils.toString(getClass().getResource("/code-block-2-result.txt").toURI(), StandardCharsets.UTF_8);

        String output = updater.updateCoordinates(input, coordinateItems);
        Assert.assertEquals(normalize(expectedOutput), normalize(output));
    }

    static String normalize(String value){
        return value
                .replace("\r", "\n")
                .replace("\n\n", "\n")
                .replace("\n\n", "\n");
    }


    @Test
    public void update_existing_code_bloc_what_it_smaller_that_new_one() throws Exception {
        String input = IOUtils.toString(getClass().getResource("/code-block-3.txt").toURI(), StandardCharsets.UTF_8);
        String expectedOutput = IOUtils.toString(getClass().getResource("/code-block-3-result.txt").toURI(), StandardCharsets.UTF_8);

        String output = updater.updateCoordinates(input, coordinateItems);
        Assert.assertEquals(expectedOutput, output);
    }

}

File: completable-reactor-graph-viewer/src/test/kotlin/ru/fix/completable/reactor/graph/viewer/code/CodeUpdaterTest.kt
package ru.fix.completable.reactor.graph.viewer.code

import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import ru.fix.completable.reactor.graph.viewer.code.CoordinateCodePhrase.*

/**
 * @author Kamil Asfandiyarov
 */
class CodeUpdaterTest {

    lateinit var codeCoordinates: List<CoordinateCodePhrase>

    fun normalize(value: String): String {
        return value
                .replace("\r", "\n")
                .replace("\n\n", "\n")
                .replace("\n\n", "\n")
    }

    @BeforeEach
    fun create_test_coordiate_items() {

        codeCoordinates = listOf(
                StartPointPhrase(500, -200),
                PlainVertexPhrase("plainVertex", 910, 901),
                ComplexVertexPhrase("complexVertex", 1170, 1270, 123, 213),
                EndPointPhrase("endPoint", 1500, 511))
    }

    private fun resourceAsText(resource: String): String {
        return javaClass.getResource(resource).readText()
    }

    @Test
    fun update_code_block_java() {
        val input = resourceAsText("/gl-code-block/code-block-1.txt")
        val expectedOutput = resourceAsText("/gl-code-block/code-block-1-result.txt")

        val output = CodeUpdater(CodeUpdater.CodeType.JAVA).updateCoordinates(input, codeCoordinates)
        assertEquals(normalize(expectedOutput), normalize(output))
    }

}